modules/opts.nix

```diff
@@ -55,31 +55,35 @@ in
   };
 
   config = {
-    extraConfigLuaPre = concatLines (
-      mapAttrsToList (
-        optionName:
-        {
-          prettyName,
-          luaVariableName,
-          luaApi,
-          ...
-        }:
-        let
-          varName = "nixvim_${luaVariableName}";
-          optionDefinitions = config.${optionName};
-        in
-        optionalString (optionDefinitions != { }) ''
-          -- Set up ${prettyName} {{{
-          do
-            local ${varName} = ${helpers.toLuaObject optionDefinitions}
+    extraConfigLuaPre =
+      let
+        content = helpers.concatNonEmptyLines (
+          mapAttrsToList (
+            optionName:
+            {
+              prettyName,
+              luaVariableName,
+              luaApi,
+              ...
+            }:
+            let
+              varName = "nixvim_${luaVariableName}";
+              optionDefinitions = config.${optionName};
+            in
+            optionalString (optionDefinitions != { }) ''
+              -- Set up ${prettyName} {{{
+              do
+                local ${varName} = ${helpers.toLuaObject optionDefinitions}
 
-            for k,v in pairs(${varName}) do
-              vim.${luaApi}[k] = v
-            end
-          end
-          -- }}}
-        ''
-      ) optionsAttrs
-    );
+                for k,v in pairs(${varName}) do
+                  vim.${luaApi}[k] = v
+                end
+              end
+              -- }}}
+            ''
+          ) optionsAttrs
+        );
+      in
+      mkIf (helpers.hasContent content) content;
```

Comment:
Is there a scenario where `content` could be empty whitespace, but not actually `""`?

I think `concatNonEmptyLines` ensures that its result either has content or is `""`.

(`concatLines` only has a trailing newline if the list is non-empty, and the list will be empty if all items failed the `hasContent` filter).

Not an issue, I'm just asking for my own understanding... In case I'm missing something again :grin:

Reply:
Your are right. My mistake. Mistake was made because because earlier I considered two options:

1. `concatLines` at the top and `helpers.hasContent content` as mkIf condition.
2. `helpers.concatNonEmptyLines` at the top and `content != ""` as a mkIf condition.

So I was rushed changing it before dinner and thought that since I'm adding something with `concatLines` at the top, I have to change also the condition, but I actually don't. With _NonEmpty_ part of the function it would work fine.

Reply:
Changed it back.

Reply:
No problem!

I hope you don't feel too much pressure to produce perfect code; there's really no such thing. :joy:

Everything can always be improved (or done differently) and sometimes it just comes down to preference or giving up and "just making it work".

Feedback in a code review is never personal and isn't always meant as criticism :smiley:

Reply:
> Everything can always be improved (or done differently) and sometimes it just comes down to preference or giving up and "just making it work".

In this case I don't agree that this is just "doing differently". For example I don't have complains in the byte-compilation PR, because I can understand that my code could be not very bad (actually I'm not a programmer, I'm sysops, and I'm junior at best at programming), or suitable, or maintainable. In this case I just think that the new behavior is wrong and counter-interact what I already fixed.

```
nix-repl> c.config.content  
"-- Ignore the user lua configuration\nvim.opt.runtimepath:remove(vim.fn.stdpath('config'))              -- ~/.config/n
vim\nvim.opt.runtimepath:remove(vim.fn.stdpath('config') .. \"/after\")  -- ~/.config/nvim/after\nvim.opt.runtimepath:r
emove(vim.fn.stdpath('data') .. \"/site\")     -- ~/.local/share/nvim/site\n\n"
```

See "\n\n" at the end? They will appear in most cases where this function is used. Fortunately stylua will strip all extra newlines, so I'm fine with that at the end. Just need to relax :).

Reply:
> `vim.opt.runtimepath:remove(vim.fn.stdpath('config'))              -- ~/.config/nvim`

Is this correct output? At a glance it looks like it's missing a linebreak :thinking: 

> In this case I don't agree that this is just "doing differently".
> See "\n\n" at the end? They will appear in most cases where this function is used.

As I said before, we can revisit in the future and figure out the best solution.

We don't need to achieve perfection in a single PR :grinning: 

> Fortunately stylua will strip all extra newlines, so I'm fine with that at the end. Just need to relax :)

:heart:

Reply:
> Is this correct output? At a glance it looks like it's missing a linebreak 🤔

Everything is fine except trailing newlines. It's added to every wrapped config here:

https://github.com/nix-community/nixvim/blob/6dc0bda459bcfb2a38cf7b6ed1d6a5d6a8105f00/modules/top-level/output.nix#L150-L153

Reply:
I think if you had additional "lines" in the `concatNonEmptyLines` call, they would also (usually) end up with `"\n\n"`, one `\n` from the line's definition and one from the concatenation.

While this would likely make the string more readable, you could debate whether or not it is "intended". So it's not just the _final_ trailing line that is an issue, although that is _more_ of an issue.

I think in another PR we can look into normalising this and figuring out exactly what we want. Unless we decide to just rely on stylua for formatting issues like this.

Reply:
> I think if you had additional "lines" in the `concatNonEmptyLines` call, they would also (usually) end up with `"\n\n"`, one `\n` from the line's definition and one from the concatenation.

Are you talking about the middle strings? That's exactly what I want: so that every string (I call it config part) is separated with empty line (empty line means double newline characters). So it is exactly what I want: double newlines between config parts and a single newline at the end. My original function was just relying on provided strings already have trailing newlines (which is the case for almost all config parts, because they are written with indented strings). So my original function worked good enough so that I didn't bother much, or would want to fix original string definition to add trailing newline for consistent behavior. So the only thing that I possibly may want to impoove is to normalize strings. Look again at my example of my «perfect» function:

```
nix-repl> lib.x86_64-linux.helpers.concatNonEmptyLinesSep "" ["a" "b\n" "" "\n" "c"]   
"a\nb\nc\n"

nix-repl> lib.x86_64-linux.helpers.concatNonEmptyLinesSep "\n" ["a" "b\n" "" "\n" "c"] 
"a\n\nb\n\nc\n"
```

Second case is what I wanted in modules/output: two newlines between config parts and only one at the end.

> I think in another PR we can look into normalising this and figuring out exactly what we want. Unless we decide to just rely on stylua for formatting issues like this.

Let's not debate on this anymore. I'm not planning making another PR. The implementation here was good enough. All that can be improved from my point of view is adding my _too magic_ function. I think it's exactly what I want most of the time and it is suitable for both cases: when I want double newlines between strings and when one newline is enough.

Stylua is great, it'll fix all trailing newlines and chomp all triple, quadruple etc newlines to double. It doesn't do anything for Vim-type configs, but who is even using it?